### ISSUE TITLE
translation of the program messages

### ISSUE THREAD

**Issue by ptakgit** [joncampbell123/dosbox-x#2488]
I started to translate the DOSBOX-X 'messages into the Czech language. But there is one problem with the code pages. In the czech Windows is used codepage 1250. I suppose, that the message.txt should be written with this (Windows) codepage, because translated messages for the menu are OK. But diacritics in messages for "PROGRAM" are wrong in the DOSBOX-X (for example :PROGRAM_CONFIG_PRIMARY_CONF), because there is used (correct ) CP852. IMHO the DOSBOX-X should translate these messages from windows codepage to the DOSBOX-X environment codepage. Is it correct idea?
![diacritics](https://user-images.githubusercontent.com/60302325/116507357-64e6cf80-a8bf-11eb-9756-96debf012432.png)

**Comment by Wengier** [comment]
Glad to hear that you have started the translation of messages. Please note that the language file the code page being saved is generally expected to match the DOS code page (inside DOSBox-X), so as a general rule if you use the code page 852 (Central Europe) in DOSBox-X, then these messages should be saved in this code page (852) as well.

There is indeed an exception: for Windows SDL1 builds, the messages for text in the drop-down menus do need to be saved in code page 1250 instead in this case, because Windows SDL1 builds use Windows's native menu in which the menu text is displayed using the Windows-only code page 1250, instead of the DOS code page inside DOSBox-X. This is considered an exception. I would suggest you to translate Czech messages for Windows SDL2 builds at start, which (like Linux builds) use internal code page instead of Windows GUI for the drop-down menus. Just save the language file using code page 852 and it should work. You can download the MinGW SDL2 build here:

* [dosbox-x-mingw64-sdl2.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64-sdl2.zip)

After that you can convert the portion of the text in the language file for all drop-down menu text from code page 852 to code page 1250 using any conversion utility, and save the whole file in a different name specifically for Windows SDL1 builds. If you look at the Spanish message translation you will find two language files below, one for general builds and one specifically for Windows SDL1 builds:

https://github.com/joncampbell123/dosbox-x/tree/master/contrib/translations/es

Hope this helps.

**Comment by joncampbell123** [comment]
That's a good point... DOSBox-X already contains guest code page to host Unicode conversion code. Perhaps the menu code should run the translated menu text through it as well, then feed the Unicode string to the WCHAR version of the Windows menu API.

**Comment by ptakgit** [comment]
Thank you, I will accept your recommendation. But - the main form left up corner windows should be converted in SDL2 form 852 to 1250 CP. In the picture below is text in messages.txt in CP1250 and are shown correctly, it is means that messges in 852 CP will be incorrect.


![MainMenu](https://user-images.githubusercontent.com/60302325/116517274-045f8e80-a8cf-11eb-9b61-f2a26a03674a.png)

**Comment by joncampbell123** [comment]
@ptakgit Windows since NT and Windows 98 are primarily Unicode based. The proposed change will not convert CP852 to CP1250, it will convert CP852 to Unicode and then feed it to the Unicode versions of the Windows API involving menus so it can display properly regardless of the host code page and language.

**Comment by joncampbell123** [comment]
Ideally on Mac OS the NSMenu can accept similar conversion from CP852 to UTF-8 Unicode.

The only odd one out is the in-window drawn SDL menu on Linux.

**Comment by ptakgit** [comment]
Unfortunatelly in SDL2 version are menu messages incorrect. The message.txt is now in CP852. Maybe the menu font is not for East Europe?
Partially translated messages.txt, CP852 messages.txt
[messages.txt](https://github.com/joncampbell123/dosbox-x/files/6397055/messages.txt)


Picture of the menu - red words are right text
![MenuSDL2](https://user-images.githubusercontent.com/60302325/116518179-3ae9d900-a8d0-11eb-8349-cc7834dd1879.png)

**Comment by Wengier** [comment]
@ptakgit You are apparently talking about the Windows system menu which will show up in both Windows SDL1 and SDL2 builds. But if DOSBox-X can convert text for Windows menu and macOS NSMenu like @joncampbell123 mentioned automatically, then this is hopefully no longer a problem.

**Comment by Wengier** [comment]
@ptakgit Try the updated MinGW builds. The update mainly addresses the Windows system menu issue since the next version is approaching.

64-bit SDL1: [dosbox-x-mingw64.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64.zip)
64-bit SDL2: [dosbox-x-mingw64-sdl2.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64-sdl2.zip)

For SDL1 build, you can try this for code page 852 and see the menu text.

``dosbox-x -lang messages.txt -set country=42,852 -set output=ttf -defaultconf``

**Comment by ptakgit** [comment]
I tested the both version but in both case the main menu is incorrect. Window menu (left up corner) is correct. Maybe the problem is wrong menu font setting - for CP 1250 it must be set with Eastern Europe charset (EASTEUROPE_CHARSET 238)?

SDL1 window menu - correct.
![SDL1WindowMenu](https://user-images.githubusercontent.com/60302325/116648163-a7baad00-a97c-11eb-86e1-9b9cec555a81.png)

SDL1 main menu - incorrect.
![SDL1MainMenu](https://user-images.githubusercontent.com/60302325/116648159-a7221680-a97c-11eb-8873-b73c3b9d8329.png)

SDL2 window menu - correct.
![SDL2WindowMenu](https://user-images.githubusercontent.com/60302325/116648168-a8ebda00-a97c-11eb-9bd1-f6c92ea2d76b.png)

SDL2 main menu - incorrect.
![SDL2MainMenu](https://user-images.githubusercontent.com/60302325/116648164-a8534380-a97c-11eb-8e7d-9ff943227f58.png)




Messages code page 1250 (partially translated)
[messages.txt](https://github.com/joncampbell123/dosbox-x/files/6402974/messages.txt)

Messages code page 852 (partially translated)
[messages852.txt](https://github.com/joncampbell123/dosbox-x/files/6402975/messages852.txt)

**Comment by Wengier** [comment]
@ptakgit As I mentioned earlier the updated builds are focusing on the fix on the Windows system menu (left up corner). Making the main drop-down menu fully Unicode compatible will be for all platforms and thus a bigger project, which will likely come later.

**Comment by ptakgit** [comment]
OK. I will temporary solve it with translation message.txt to the text without diacritics.

**Comment by Wengier** [comment]
@ptakgit You can do that of course, but you can also make another language file with diacritics as normal, which will eventually be supported on all platforms.

**Comment by ptakgit** [comment]
Last question - is in the DOSBOX-X version you sended me ebove included also the modification https://github.com/joncampbell123/dosbox-x/commit/427fd343cd33ec437e20d3dbd004ca1484f118d0 for the faster file copying? I tested copying now with the same long time result.

**Comment by Wengier** [comment]
@ptakgit Yes, it does. But please, use the default setting (instead of your own config file) for the testing. For example, I already mentioned above that for SDL1 build, you can try this for code page 852:

```
dosbox-x -lang messages852.txt -set country=42,852 -set output=ttf -defaultconf
```

The result may be different from using your own config file.

**Comment by Wengier** [comment]
@ptakgit Please try the updated 64-bit MinGW build. With your messages852.txt file both the main drop-down menu and the Windows system menu should look fine with the correct keyboard layout (e.g. keyboardlayout=cz):

* [dosbox-x-mingw64.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64.zip)

Screenshot:

![image](https://user-images.githubusercontent.com/8216923/116852326-1e94b780-abc2-11eb-9616-219597580be5.png)

So you should be able to translate text with diacritics normally already.

**Comment by ptakgit** [comment]
Some items are OK, other are wrong - see the picture.


![Menu03-05](https://user-images.githubusercontent.com/60302325/116875560-698ee900-ac1b-11eb-8215-b99f0b95064d.png)

**Comment by ptakgit** [comment]
I found the reason of the wrong texts. CodePage wasn't correct. For setting country and codepage doesn't works option [config] country=42,852 in .conf file! Only country has been setted on my DOSBOX-X, not codepage! It could be solved by the command chcp 852 in the [autoexec] section, but the country option doesn't works compatible with DOS. Second correct possibility is to start DOSBOX-X with parameters -set country=42,852, but using country= ... in the config should be possible also.

In conclusion - the translation of the menu now probably works, so I will continue with the translation into Czech. But please also look at the code page settings in the .conf file.

**Comment by Wengier** [comment]
@ptakgit Just let you know the main factor for the current version for the local switch is the KEYB command (or keyboardlayout config option), not the country config option. For the country code, you can also separately set it using COUNTRY command in the DOS shell.

**Comment by Wengier** [comment]
@@ptakgit Please check out the updated 64-bit MinGW SDL1 build. It now supports saving to and loading from UTF-8 encoded language files, so I think you can just save your language file in UTF-8 instead of local code pages. Also, you can just set country=42,852 for this as before.

* [dosbox-x-mingw64.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64.zip)

The welcome banner messages have also been reorganized so that translating them from stretch should be easier than before (and more compatible with other code pages too), but for the existing language file please copy the old translated text for the welcome banner messages manually.

**Comment by ptakgit** [comment]
I translated only part of the menu messagess to now. But translated menu texts are now correct in 852CP and UTF-8. If I will found any problem with menu, I will send a report.
Thank you.

**Comment by Wengier** [comment]
@ptakgit Just let you know that you can now also use the Windows SDL2 build for the menu translation, which should support language files in local code pages and UTF-8 too. It is always safer to set both `country` and `keyboardlayout` options (e.g. `country=42,852` and `keyboardlayout=cz`) in the config file. The MinGW build is available from:

* [dosbox-x-mingw64-sdl2.zip](http://cndos.fam.cx/dosware/dosbox-x-mingw64-sdl2.zip)

**Comment by ptakgit** [comment]
SDL2 is also OK.

**Comment by Wengier** [comment]
@ptakgit We have made some small improvements to the language files following the discussion in #2326 in the last a few days. The updated 32-bit Visual Studio builds are available below:

* Visual Studio SDL1 build: [dosbox-x-update.zip](http://cndos.fam.cx/dosware/dosbox-x-update.zip)
* Visual Studio SDL2 build: [dosbox-x-sdl2-update.zip](http://cndos.fam.cx/dosware/dosbox-x-sdl2-update.zip)

**Comment by Wengier** [comment]
@ptakgit I have been using your (partially-translated) Czech language file for verifying the multi-language functionality of different types of menus (including the SDL-drawn menu), and it has indeed helped a lot. Please let me know if there is any update on the Czech language file. Thanks!

**Comment by ptakgit** [comment]
I've been busy lately, I won't be able to continue the translation until about July. After I finish it,  I will send it.   I downloaded the June version and it looks good. Thanks.
